File: example/real/arm_traj_ctrl.cpp
/**
 * arm_traj_ctrl — 实机测地线轨迹交互控制
 *
 * 终端输入末端目标 "x y z [roll pitch yaw(rad)]"，确认后执行；Ctrl+C 回零退出。
 *
 * 使用: sudo ./arm_traj_ctrl [串口] [YAML] [URDF]
 *
 * 轨迹规划算法（se3Geodesic 测地线插值 + CLIK 跟踪）：
 *   - 测地线：SE(3) 空间最短路径插值，姿态平滑
 *   - CLIK：阻尼最小二乘逆运动学，带零空间关节限位规避
 */

#include "application/arm_controller.h"

#include <iostream>
#include <sstream>
#include <string>
#include <chrono>

using rebot::ArmController;
using namespace std::chrono;

static ArmController arm_ctrl;

static void print_pose(const char* prefix, const pinocchio::SE3& T)
{
    const auto&     p   = T.translation();
    Eigen::Vector3d rpy = T.rotation().eulerAngles(2, 1, 0).reverse();
    printf("%s  pos[%.3f %.3f %.3f]  rpy[%.3f %.3f %.3f]\n",
           prefix, p[0], p[1], p[2], rpy[0], rpy[1], rpy[2]);
}

int main(int argc, char* argv[])
{
    const std::string dev  = (argc > 1) ? argv[1] : "/dev/ttyACM0";
    const std::string yaml = (argc > 2) ? argv[2] : "";
    const std::string urdf = (argc > 3) ? argv[3] : "";

    if (!arm_ctrl.init(dev, yaml, urdf)) return 1;

    printf("输入: x y z [roll pitch yaw(rad)]  Ctrl+C 退出\n\n");

    std::string line;
    while (arm_ctrl.running()) {
        print_pose("当前", arm_ctrl.fk());

        if (!std::getline(std::cin, line) || !arm_ctrl.running()) break;
        if (line.find_first_not_of(" \t") == std::string::npos) continue;

        std::istringstream iss(line);
        double x, y, z, ro = 0, pi = 0, ya = 0;
        if (!(iss >> x >> y >> z)) {
            fprintf(stderr, "格式错误: x y z [roll pitch yaw]\n");
            continue;
        }
        iss >> ro >> pi >> ya;

        printf("目标  pos[%.3f %.3f %.3f]  rpy[%.3f %.3f %.3f]\n是否执行? [y/N]: ",
               x, y, z, ro, pi, ya);
        fflush(stdout);
        std::string yn;
        if (!std::getline(std::cin, yn)) break;
        if (yn.empty() || (yn[0] != 'y' && yn[0] != 'Y')) { printf("已取消\n\n"); continue; }

        // 测地线轨迹规划 + CLIK 跟踪
        auto start_time = high_resolution_clock::now();
        if (!arm_ctrl.move_to_geodesic(ArmController::pose(x, y, z, ro, pi, ya))) {
            fprintf(stderr, "[错误] IK 或轨迹规划失败\n\n");
            continue;
        }
        auto end_time = high_resolution_clock::now();
        double duration = duration_cast<milliseconds>(end_time - start_time).count() / 1000.0;

        print_pose("到达", arm_ctrl.fk());
        printf("[轨迹] 耗时: %.2f 秒\n\n", duration);
    }

    arm_ctrl.shutdown();
    return 0;
}

File: example_py/arm_traj_ctrl.py
#!/usr/bin/env python3
"""
arm_traj_ctrl.py — 实机测地线轨迹交互控制（Python 版）

终端输入末端目标 "x y z [roll pitch yaw(rad)]"，确认后执行；Ctrl+C 回零退出。
使用: sudo python3 arm_traj_ctrl.py [串口] [YAML] [URDF]

等价 C++ 版: example/real/arm_traj_ctrl.cpp
"""

import sys
import os
import math
import numpy as np

sys.path.insert(0, "../build")
import rebot_py

# 自动计算项目根目录（基于脚本自身位置）
_script_dir = os.path.dirname(os.path.abspath(__file__))
_project_root = os.path.dirname(_script_dir)  # example_py 的上一级


def se3_to_pos_rpy(T: np.ndarray):
    """从 4×4 位姿矩阵提取 (pos, rpy)"""
    pos = T[:3, 3]
    R = T[:3, :3]
    sy = math.sqrt(R[0, 0] ** 2 + R[1, 0] ** 2)
    if sy > 1e-6:
        roll  = math.atan2(R[2, 1], R[2, 2])
        pitch = math.atan2(-R[2, 0], sy)
        yaw   = math.atan2(R[1, 0], R[0, 0])
    else:
        roll  = math.atan2(-R[1, 2], R[1, 1])
        pitch = math.atan2(-R[2, 0], sy)
        yaw   = 0.0
    return pos, np.array([roll, pitch, yaw])


def print_pose(prefix: str, T: np.ndarray):
    pos, rpy = se3_to_pos_rpy(T)
    print(f"{prefix}  pos[{pos[0]:.3f} {pos[1]:.3f} {pos[2]:.3f}]"
          f"  rpy[{rpy[0]:.3f} {rpy[1]:.3f} {rpy[2]:.3f}]")


def main():
    # 如果用户没有传入参数，使用自动计算的路径
    dev  = sys.argv[1] if len(sys.argv) > 1 else "/dev/ttyACM0"
    yaml = sys.argv[2] if len(sys.argv) > 2 else os.path.join(_project_root, "config/arm.yaml")
    urdf = sys.argv[3] if len(sys.argv) > 3 else os.path.join(
        _project_root, "urdf/reBot-DevArm_fixend_description/urdf/reBot-DevArm_fixend.urdf")

    arm = rebot_py.ArmController()
    if not arm.init(dev, yaml, urdf):
        sys.exit(1)

    print("输入: x y z [roll pitch yaw(rad)]  Ctrl+C 退出\n")

    try:
        while arm.running():
            print_pose("当前", arm.fk())

            line = input("> ").strip()
            if not line:
                continue

            parts = line.split()
            if len(parts) < 3:
                print("格式错误: x y z [roll pitch yaw]")
                continue

            x, y, z = float(parts[0]), float(parts[1]), float(parts[2])
            ro = float(parts[3]) if len(parts) > 3 else 0.0
            pi = float(parts[4]) if len(parts) > 4 else 0.0
            ya = float(parts[5]) if len(parts) > 5 else 0.0

            print(f"目标  pos[{x:.3f} {y:.3f} {z:.3f}]  rpy[{ro:.3f} {pi:.3f} {ya:.3f}]")
            yn = input("是否执行? [y/N]: ").strip()
            if not yn or yn[0].lower() != "y":
                print("已取消\n")
                continue

            target = rebot_py.ArmController.pose(x, y, z, ro, pi, ya)
            if not arm.move_to_geodesic(target):
                print("IK 或规划失败\n")
                continue

            arm.refresh_hw()
            print_pose("到达", arm.fk())
            print()

    except (KeyboardInterrupt, EOFError):
        pass

    arm.shutdown()


if __name__ == "__main__":
    main()
